Using-GTSAM-EXPORT.md

```diff
@@ -0,0 +1,41 @@
+```
+
+Sorry for the long error message there, but I want to quickly decode it to help understand what is going on.  First, there is an unresolved symbol `gtsam::SO3::print`.  This can occur because _either_ `GTSAM_EXPORT` was not added to the print function definition when it should have been, _OR_ because `GTSAM_EXPORT` was added to the print function definition when it is fully declared in the header.  This error was detected while compiling `check_geometry_program` and pulling in `...\testSO3.obj`.  Specifically, within the function call `gtsam::Testabl<class gtsam::SO3>::Print (...)`.  Note that this error did _not_ occur when compiling the library that actually has SO3 inside of it.
+
+## But Why?
+To me, the behavior of GTSAM_EXPORT was very baffling.  However, I believe it can be explained by understanding the rules that MSVC operates under.  
+
+First, we need to understand exactly what `GTSAM_EXPORT` does.  When you use `GTSAM_EXPORT` on a function (or class) definition, it actually does two things.  First, when you are building the library, it turns into a "dllexport" command, telling the compiler that this function should go into the DLL and be visible externally.  When you pull in that same header for a different library, it switches to a "dllimport" command, telling the compiler it should find this function in a DLL somewhere.  This leads to the first rule the compiler uses.  (Note that I say "compiler rules" when the rules may actually be in the linker, but I am conflating the two terms here when I speak of the "compiler rules".) 
+
+***Compiler Rule #1*** If a `dllimport` command is used in defining a function or class, that function or class _must_ be found in a DLL.
+
+Rule #1 doesn't seem very bad, until you combine it with rule #2
+
+***Compiler Rule #2*** Anything defined in a header file is not included in a DLL.
+
+When these two rules are combined, you get some very confusing results.  For example, a class which is completely defined in a header (e.g. LieMatrix) cannot use `GTSAM_EXPORT` in its definition.  If LieMatrix is defined with `GTSAM_EXPORT`, then the compiler _must_ find LieMatrix in a DLL.  Because LieMatrix is a header-only class, however, it can't find it, leading to a very confusing "I can't find this symbol" type of error.  Note that the linker says it can't find the symbol even though the compiler found the header file that completely defines the class.
+
+This leads to another compiler rule that finishes explaining the usage rules proposed above.
+
+***Compiler Rule #3*** If a class is defined with `GTSAM_EXPORT`, then any classes it inherits from must be definable with a dllexport command as well.
```

Comment:
Again, I'm not 100% sure of this... (but may be wrong!). 
For example, [this MSVC warning](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4275?view=vs-2019), which I found many times, talks specifically about that situation, and it's not forbidden, it "only" **could** lead to data corruption. However, I have been using DLL-exported classes that inheriting from header-only classes for years, without problems. Nowadays, I avoid it just to be safe, but I wouldn't say it's carved in stone that it cannot be done...

Reply:
I'm not 100% sure. I know that for the "Point?" classes in gtsam/geometry, because they derive from Eigen classes, DLLimport problems were popping up until I moved the GTSAM_EXPORT inside the class to the individual functions.  It may be that it violates on the conditions discussed at the [MSVC warning](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4275?view=vs-2019) page you kindly pointed towards.

In the meantime, I will weaken the verbiage to say that things could go bad and it is generally not recommended.